docs/tutorialkit.dev/src/content/docs/guides/using-low-level-apis.mdx

```diff
@@ -0,0 +1,189 @@
+```
+
+### Example usage in React component
+
+Here is an example of custom React component that reads contents of `index.js` and gives user hints about the expected content.
+
+<HelpButton client:load />
+
+```tsx title="src/components/HelpButton.tsx"
+import tutorialStore from "tutorialkit:store";
+import { Dialog } from "@tutorialkit/react";
+import { useState } from "react";
+
+export function HelpButton() {
+  const [message, setMessage] = useState<string | null>(null);
+
+  function onClick() {
+    const files = tutorialStore.documents.value;
+    const index = files["/index.js"].value as string;
+
+    if (index.length === 0) {
+      setMessage("Your index.js should not be empty");
+    } else if (!index.includes("export default")) {
+      setMessage("Your index.js should have a default export");
+    } else {
+      setMessage("All good");
+    }
+  }
+
+  return (
+    <>
+      <button
+        onClick={onClick}
+        className="px-4 py-1 rounded-md bg-tk-elements-primaryButton-backgroundColor text-tk-elements-primaryButton-textColor"
+      >
+        Help
+      </button>
+
+      {message && (
+        <Dialog title="Help" confirmText="OK" onClose={() => setMessage(null)}>
+          {message}
+        </Dialog>
+      )}
+    </>
+  );
+}
+```
+
+```mdx title="content.mdx"
+---
+type: lesson
+title: Lesson Topic
+---
+
+import { HelpButton } from "@/components/HelpButton";
+
+# Lesson Topic
+
+In this lesson we'll learn about...
+Click this button if you need help: <HelpButton client:load />
+```
+
+## Tutorial Core
+
+You can access Tutorial Core by importing the `tutorialkit:core` entrypoint.
+
+```ts
+import * as core from "tutorialkit:core";
+```
+
+The Tutorial Core provides access to webcontainer APIs. You can use it for example to read and modify files on the virtual file system.
+
+```ts
+import * as core from "tutorialkit:core";
+
+const webcontainer = await core.webcontainer;
+const contents = await webcontainer.fs.readdir("/", "utf8");
+
+console.log(contents);
+[
+  "index.html",
+  "node_modules",
+  "package-lock.json",
+  "package.json",
+  "vite.config.ts"
+]
+```
+
+### Example usage in React component
+
+Here is an example of custom React component that looks for updates in `/index.js` and provides user information about their lesson status.
+
+<video src={taskListVideo} controls />
+
+```tsx title="src/components/TaskList.tsx"
+import * as core from "tutorialkit:core";
+import { useEffect, useState } from "react";
+
+export function TaskList() {
+  const [hasDefaultExport, setDefaultExport] = useState<boolean>(false);
+  const [hasSumFunction, setSumFunction] = useState<boolean>(false);
+
+  useEffect(() => {
+    const unsubscribe = subscribeFileChanges("/index.js", (content) => {
+      // Naive text based checking just for example. Use proper AST analysis in real lessons.
+      setDefaultExport(content.includes("export default"));
+      setSumFunction(content.includes("function sum"));
```

Comment:
This is a pretty cool idea but I'm a bit torn on this one.

On one hand, I'm not sure if it makes sense for us to recommend doing any form of pattern matching against an AST (especially in the UI thread).

But the other solution I have in mind, which is to run a program in webcontainer that import the file and run a test against it, and check the exit code of that process, is not going to be without its own issues too.

I suppose we should probably have example of both.

Reply:
Yeah you're right I forgot for a minute that  those files should be pretty small 👍

Reply:
Now using `magicast` for this. https://stackblitz.com/edit/tutorialkit-api-use-helpbutton?file=src/components/HelpButton.tsx&file=src/content/tutorial/1-basics/1-introduction/1-welcome/content.mdx

Reply:
This is soooo cool! :heart_eyes: Good job on this! :smiley: